src/equations/shallow_water_two_layer_1d.jl

```diff
@@ -0,0 +1,594 @@
+  v1_rr, v2_rr = velocity(u_rr, equations)
+
+  # Average each factor of products in flux
+  v1_avg = 0.5 * (v1_ll + v1_rr)
+  v2_avg = 0.5 * (v2_ll + v2_rr)
+  p1_avg = 0.5 * equations.gravity * h_upper_ll * h_upper_rr
+  p2_avg = 0.5 * equations.gravity * h_lower_ll * h_lower_rr
+
+  # Calculate fluxes
+  f1 = 0.5 * (h_v1_upper_ll + h_v1_upper_rr)
+  f2 = f1 * v1_avg + p1_avg
+  f3 = 0.5 * (h_v2_lower_ll + h_v2_lower_rr)
+  f4 = f3 * v2_avg + p2_avg
+
+  return SVector(f1, f2, f3, f4, zero(eltype(u_ll)))
+end
+
+
+"""
+    flux_es_fjordholm_etal(u_ll, u_rr, orientation,
+                           equations::ShallowWaterTwoLayerEquations1D)
+
+Entropy stable surface flux for the two-layer shallow water equations. Uses the entropy 
+conservative flux_fjordholm_etal and adds a Lax-Friedrichs type dissipation dependent on the jump 
+of entropy variables.
+
+Further details are available in the paper:
+- Ulrik Skre Fjordholm (2012)
+  Energy conservative and stable schemes for the two-layer shallow water equations.
+  [DOI: 10.1142/9789814417099_0039](https://doi.org/10.1142/9789814417099_0039)
+It should be noted that the equations are ordered differently and the
+designation of the upper and lower layer has been changed which leads to a slightly different
+formulation.
+"""
+@inline function flux_es_fjordholm_etal(u_ll, u_rr,
+                                        orientation::Integer,
+                                        equations::ShallowWaterTwoLayerEquations1D)
+  # Compute entropy conservative flux but without the bottom topography
+  f_ec = flux_fjordholm_etal(u_ll, u_rr,
+                              orientation,
+                              equations)[1:4]
+
+  # Get maximum signal velocity
+  λ = max_abs_speed_naive(u_ll, u_rr, orientation, equations)
+
+  # Get entropy variables but without the bottom topography
+  q_rr = cons2entropy(u_rr,equations)[1:4]
+  q_ll = cons2entropy(u_ll,equations)[1:4]
+
+  # Average values from left and right
+  u_avg = (u_ll + u_rr) / 2
+
+  # Introduce variables for better readability
+  rho_upper = equations.rho_upper
+  rho_lower = equations.rho_lower
+  g    = equations.gravity
+  drho = rho_upper - rho_lower
+
+  # Entropy Jacobian matrix
+  H = MArray{Tuple{4,4}, eltype(u_ll), 2}(undef)
+  # Use symmetry properties to compute matrix
+  H[:,1] = 
+    [-rho_lower/(g*rho_upper*drho);;
+    -rho_lower*u_avg[2]/(g*rho_upper*u_avg[1]*drho);;
+      1.0/(g*drho);;
+      u_avg[4]/(g*u_avg[3]*drho)]
+  H[:,2] = 
+    [H[2,1];;
+      (g*rho_upper*u_avg[1]^3 - g*rho_lower*u_avg[1]^3 - rho_lower*u_avg[2]^2)/(g*rho_upper*u_avg[1]^2*drho);;
+      u_avg[2]/(g*u_avg[1]*drho);;
+      u_avg[2]*u_avg[4]/(g*u_avg[1]*u_avg[3]*drho)]
+  H[:,3] =
+    [H[3,1];;
+      H[3,2];;
+    -1.0/(g*drho);;
+    -u_avg[4]/(g*u_avg[3]*drho)]
+  H[:,4] = 
+    [H[4,1];;
+      H[4,2];;
+      H[4,3];;
+      (g*rho_upper*u_avg[3]^3 - g*rho_lower*u_avg[3]^3 - rho_lower*u_avg[4]^2)/(g*rho_lower*u_avg[3]^2*drho)]
```

Comment:
Allocations

Reply:
You could use SVector or preallocate the array

Reply:
Or you could try using an `MMatrix` - if it's clear that it won't leave scope, newer versions of Julia might not need to allocate memory on the heap for it.

Reply:
Isn't this what I am currently doing as I am assigning the matrix `H`, as `MArray`?

Reply:
Thanks for the explanation! I can get rid of the allocations with an `SMatrix` now.

However, I also noticed that using a slice of a vector as in this example:
`q_ll = cons2entropy(u_ll,equations)[1:4]`
leads to further allocations. Do you have any ideas how to do this without allocations?

Reply:
That's a bit ugly. You could of course just create a new `SVector` with the first four components of the return value of `cons2entropy`. Maybe there is a nice/short way, but I am on my phone right now and can't search the docs nicely

Reply:
Did you benchmark it to ensure everything works as intended?

Reply:
Yes, I have checked the allocations with BenchmarkTools.jl.

Reply:
Thanks!